webknossos/webknossos/dataset/_utils/pims_tiff_reader.py

```diff
@@ -0,0 +1,116 @@
+import tempfile
+from os import PathLike
+from pathlib import Path
+from typing import Set, Tuple
+
+import numpy as np
+
+try:
+    from pims import FramesSequenceND
+except ImportError as e:
+    raise RuntimeError(
+        "Cannot import pims, please install it e.g. using 'webknossos[all]'"
+    ) from e
+
+try:
+    import tifffile
+except ImportError as e:
+    raise RuntimeError(
+        "Cannot import tifffile, please install it e.g. using 'webknossos[tifffile]'"
+    ) from e
+
+
+class PimsTiffReader(FramesSequenceND):
+    @classmethod
+    def class_exts(cls) -> Set[str]:
+        return {"tif", "tiff"}
+
+    # class_priority is used in pims to pick the reader with the highest priority.
+    # We decided to use a custom reader for tiff files to support images with more than 3 dimensions out of the box.
+    # Default is 10, and bioformats priority is 2.
+    # Our custom reader for imagej_tiff has priority 20.
+    # See http://soft-matter.github.io/pims/v0.6.1/custom_readers.html#plugging-into-pims-s-open-function
+    class_priority = 19
+
+    def __init__(self, path: PathLike) -> None:
+        super().__init__()
+
+        path = Path(path)
+
+        _tiff = tifffile.TiffFile(path).series[0]
+        self._tiff_axes = tuple(_tiff.axes.lower())
+        for axis, shape in zip(self._tiff_axes, _tiff.shape):
+            self._init_axis(axis, shape)
+
+        self._tiff_shape = _tiff.shape
+
+        # Selecting the first page to get the dtype and shape
+        if hasattr(_tiff, "pages"):
+            _tmp = _tiff.pages[0]
+        else:
+            _tmp = _tiff["pages"][0]
+        assert _tmp is not None, "No pages found in tiff file."
+        self._dtype = _tmp.dtype or np.dtype("uint8")
+        self._shape = _tmp.shape
+        with tempfile.TemporaryFile() as tmp:
+            self._memmap = np.memmap(
+                tmp,
+                dtype=self._dtype,
+                shape=_tiff.shape,
+                mode="w+",
+            )
```

Comment:
Why is the memmap in a tmp file? Isn't that creating a copy of the image?

Reply:
Yes, I tried to use the tiff file memmap function to don't create a copy, but it only works for a few tiff files. OME tiffs as our example data for a 4d image are not memory-mappable, according to the tifffile library.

Reply:
Sure, I can catch the error thrown be the tifffile library and use my code as a fallback.